include/configoptions.h

```diff
@@ -0,0 +1,310 @@
+        */
+        ConfigOptions(ConfigOptions&& cfgopts) = default;
+
+        /*!
+        *   \brief ConfigOptions move assignment operator
+        *   \param cfgopts The ConfigOptions to move and assign
+        */
+        ConfigOptions& operator=(ConfigOptions&& cfgopts) = default;
+
+        /*!
+        *   \brief ConfigOptions destructor
+        */
+        virtual ~ConfigOptions() = default;
+
+        /////////////////////////////////////////////////////////////
+        // Factory construction methods
+
+        /*!
+        *   \brief Instantiate ConfigOptions, getting selections from
+        *          environment variables. If \p db_prefix is non-empty,
+        *          then "{db_prefix}_" will be prepended to the name of
+        *          each environment variable that is read.
+        *   \param db_prefix The prefix to use with environment variables,
+        *                    or an empty string to disable prefixing
+        *   \returns The constructed ConfigOptions object
+        *   \throw SmartRedis::Exception if db_prefix contains invalid
+        *          characters
+        */
+        static std::unique_ptr<ConfigOptions> create_from_environment(
+            const std::string& db_prefix);
+
+// Configuration via JSON file or JSON blob is anticipated in the future
+// but not supported yet
+#ifdef FUTURE_CONFIG_SUPPORT
+        /*!
+        *   \brief Instantiate ConfigOptions, getting selections from
+        *          a file with JSON data.
+        *   \param filename A UTF-8 file with JSON data containing the
+        *                   configuration data
+        *   \returns The constructed ConfigOptions object
+        *   \throw SmartRedis::Exception if the file cannot be found or
+        *          the data within it cannot be parsed
+        */
+        static std::unique_ptr<ConfigOptions> create_from_file(
+            const std::string& filename);
+
+        /*!
+        *   \brief Instantiate ConfigOptions, getting selections from
+        *          a string containing a JSON blob
+        *   \param json_blob A JSON blob containing the configuration data
+        *   \returns The constructed ConfigOptions object
+        *   \throw SmartRedis::Exception if the JSON blob cannot be parsed
+        */
+        static std::unique_ptr<ConfigOptions> create_from_string(
+            const std::string& json_blob);
+#endif
+
+        /////////////////////////////////////////////////////////////
+        // Option access
+
+        /*!
+        *   \brief Retrieve the value of a numeric configuration option
+        *          from the selected source
+        *   \param key The name of the configuration option to retrieve
+        *   \returns The value of the selected option
+        *   \throw Throws SRKeyException if the option was not set in the
+        *          selected source
+        */
+        int64_t get_integer_option(const std::string& key);
```

Comment:
Could we call this something other than `key`? Maybe `option_name`

Reply:
done for all client languages